src/Features/Core/Portable/CodeRefactorings/CodeRefactoringContextExtensions.cs

```diff
@@ -40,32 +40,67 @@ internal static class CodeRefactoringContextExtensions
+        return TryGetRelevantNode<TSyntaxNode>(parsedDocument, context.Span, allowEmptyNode, context.CancellationToken);
+    }
 
     public static Task<ImmutableArray<TSyntaxNode>> GetRelevantNodesAsync<TSyntaxNode>(this CodeRefactoringContext context) where TSyntaxNode : SyntaxNode
         => GetRelevantNodesAsync<TSyntaxNode>(context, allowEmptyNodes: false);
 
-    public static Task<ImmutableArray<TSyntaxNode>> GetRelevantNodesAsync<TSyntaxNode>(this CodeRefactoringContext context, bool allowEmptyNodes) where TSyntaxNode : SyntaxNode
-        => GetRelevantNodesAsync<TSyntaxNode>(context.Document, context.Span, allowEmptyNodes, context.CancellationToken);
+    public static async Task<ImmutableArray<TSyntaxNode>> GetRelevantNodesAsync<TSyntaxNode>(this CodeRefactoringContext context, bool allowEmptyNodes) where TSyntaxNode : SyntaxNode
+    {
+        var parsedDocument = await ParsedDocument.CreateAsync(context.Document, context.CancellationToken).ConfigureAwait(false);
+        return GetRelevantNodes<TSyntaxNode>(parsedDocument, context.Span, allowEmptyNodes, context.CancellationToken);
+    }
 
-    public static Task<TSyntaxNode?> TryGetRelevantNodeAsync<TSyntaxNode>(this Document document, TextSpan span, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
-        => TryGetRelevantNodeAsync<TSyntaxNode>(document, span, allowEmptyNode: false, cancellationToken);
+    public static async Task<TSyntaxNode?> TryGetRelevantNodeAsync<TSyntaxNode>(this Document document, TextSpan span, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+    {
+        var parsedDocument = await ParsedDocument.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+        return TryGetRelevantNode<TSyntaxNode>(parsedDocument, span, cancellationToken);
+    }
+
+    public static TSyntaxNode? TryGetRelevantNode<TSyntaxNode>(this ParsedDocument document, TextSpan span, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+        => TryGetRelevantNode<TSyntaxNode>(document, span, allowEmptyNode: false, cancellationToken);
 
-    public static async Task<TSyntaxNode?> TryGetRelevantNodeAsync<TSyntaxNode>(this Document document, TextSpan span, bool allowEmptyNode, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+    public static TSyntaxNode? TryGetRelevantNode<TSyntaxNode>(this ParsedDocument document, TextSpan span, bool allowEmptyNode, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
     {
-        var potentialNodes = await GetRelevantNodesAsync<TSyntaxNode>(document, span, allowEmptyNode, cancellationToken).ConfigureAwait(false);
-        return potentialNodes.FirstOrDefault();
+        using var result = TemporaryArray<TSyntaxNode>.Empty;
+        AddRelevantNodes(document, span, allowEmptyNode, stopOnFirst: true, ref result.AsRef(), cancellationToken);
+        return result.FirstOrDefault();
```

Comment:
this helper (used in a ton of places) can now stop on the first result, and not do more work walking the tree.